### PR TITLE
Remove dependency on goldfinger gem

### LINKED ISSUE #14091: Webfinger timeout seems too short : increasing sidekiq error rate caused by PR #13836
### Expected behaviour

Acceptable sidekiq error rate.

### Actual behaviour

Sidekiq retry and dead job queue growing more than before.

Below is error message sample:

> HTTP::TimeoutError
> execution expired
> /usr/home/mastodon/live/app/helpers/webfinger_helper.rb:8:in `initialize'
> /usr/home/mastodon/live/app/helpers/webfinger_helper.rb:8:in `open'
> /usr/home/mastodon/live/app/helpers/webfinger_helper.rb:8:in `block in connect'

### Steps to reproduce the problem

Upgrade mastodon to master branch d890abfcab076f5df4eafb75397435d32c641e20 (PR 13836) or newer.

### Specifications

This problem is caused by PR #13836

### Amendment Proposal

I think connection timeout is too short for busy servers.
It's difficult to find optimal value of timeout, but, based on my investigation, It seems 120sec. is good start for us.

line 29 of app/helpers/webfinger_helper.rb:

```ruby
timeout_options: {
 write_timeout: 10,
 connect_timeout: 120,  # <== increase timeout value
 read_timeout: 10,
},
```

### Note

This problem seems to occur while processing ActivityPub Announce activity (a.k,a, Boost, BT).
I haven't checked in detail, but there may be performance issues in processing Announce activity.

### PR DESCRIPTION
**Pull request by Gargron**
There are edge cases where requests to certain hosts timeout when using the vanilla HTTP.rb gem, which the goldfinger gem uses. Now that we no longer need to support OStatus servers, webfinger logic is so simple that there is no point encapsulating it in a gem, so we can just use our own Request class. With that, we benefit from more robust timeout code and IPv4/IPv6 resolution.

Fix #14091

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
> Some of our previous documentation has recommended people to use the host-meta path to setup domain->subdomain forwarding for Mastodon, but I believe that it could be done better by using the webfinger path directly.

Does that mean this isn't implemented yet? How would domain to subdomain forwarding work?

**Comment by Gargron** [comment]
It works by proxying webfinger responses from the subdomain through the top-level domain. So you add a location block to your top level domain's nginx config specifically for the /.well-known/host-meta path that uses proxy_pass to Mastodon. Basically that documentation was bad advice. Instead of using the host-meta path, people should've been proxying the webfinger path itself.

**Comment by ThibG** [comment]
`host-meta` isn't unspecified, it's in https://www.rfc-editor.org/rfc/rfc6415.txt, though it doesn't appear to be in use in the Webfinger specification (apparently, it was in earlier drafts but eventually dropped).

I'm not opposed to dropping this, but I am indeed worried about instances that may currently rely on it.

**Comment by Gargron** [comment]
I didn't mean to say that host-meta wasn't *a* spec, just that it isn't in the webfinger one, and the only reason we have it right now is because GNU social relied on host-meta to work around various PHP path rewriting woes.

**Comment by Gargron** [comment]
I re-added host-meta logic.

File: app/helpers/webfinger_helper.rb [hunk 1]
```diff
@@ -1,38 +1,7 @@
...
-      @socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1) if nodelay
-    end
-  end
-end
-
 module WebfingerHelper
   def webfinger!(uri)
-    hidden_service_uri = /\.(onion|i2p)(:\d+)?$/.match(uri)
-
-    raise Mastodon::HostValidationError, 'Instance does not support hidden service connections' if !Rails.configuration.x.access_to_hidden_service && hidden_service_uri
```
Comment (ThibG) [inline app/helpers/webfinger_helper.rb:18]: Why remove this part?

File: app/lib/webfinger.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  def initialize(uri)
+    @uri    = uri
+    @domain = uri.split('@').last
+  end
+
+  def perform
+    Response.new(body_from_webfinger)
+  rescue Oj::ParseError
+    raise Webfinger::Error, "Invalid JSON in response for #{@uri}"
```
Comment (ThibG) [inline app/lib/webfinger.rb:34]: You should also rescue `Addressable::URI::InvalidURIError` and `Nokogiri::XML::XPath::SyntaxError`

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
In addition to the inlined comments, some tests need to be adapted to changes in `WebfingerHelper`'s interface.

Also, you did not mention that you dropped support for XML-based webfinger replies. They're not in the spec and that simplifies the code, so I'm ok with that in principle, but it'd be better if we checked there aren't any implementation left that serves XML only.

**Comment by ThibG** [comment]
For now I have surveyed about 5k of the 44k accounts known to my instance as AP-compatible accounts. The results so far are:
- 9 instances would be affected by us dropping support for host-meta
- 1 instance is affected by the host-meta fallback rules changing (they return 410 instead of 404 for the first webfinger query)
- none are affected by dropping support for XML to the webfinger query

File: app/helpers/webfinger_helper.rb (hunk 1 above)
Comment (Gargron) [reply app/helpers/webfinger_helper.rb:18]: Because Request handles that

File: app/services/resolve_account_service.rb [hunk 3]
```diff
@@ -26,11 +26,10 @@ def call(uri, options = {})
 
     @account ||= Account.find_remote(@username, @domain)
 
-    return @account if @account&.local? || !webfinger_update_due?
+    return @account if @account&.local? || @domain.nil? || !webfinger_update_due?
```
Comment (ThibG) [inline app/services/resolve_account_service.rb:29]: Is there a case where `@account.local?` would be false and `@domain.nil?` would be true?

File: app/views/well_known/host_meta/show.xml.ruby [hunk 4]
```diff
@@ -5,7 +5,7 @@ doc << Ox::Element.new('XRD').tap do |xrd|
 
   xrd << Ox::Element.new('Link').tap do |link|
     link['rel']      = 'lrdd'
-    link['type']     = 'application/xrd+xml'
+    link['type']     = 'application/jrd+json'
```
Comment (ThibG) [inline app/views/well_known/host_meta/show.xml.ruby:8]: That would mean breaking compatibility with implementations which only handle XRD for LLRD (I don't know if there is any). Maybe we could advertise both.

File: app/services/resolve_account_service.rb (hunk 3 above)
Comment (Gargron) [reply app/services/resolve_account_service.rb:29]: There would be a case where @account is nil *and* @domain is nil. It looks like before this, doing a local mention of a non-existing account attempted to perform a webfinger resolution (which failed due to invalid URI structure).

File: app/views/well_known/host_meta/show.xml.ruby (hunk 4 above)
Comment (Gargron) [reply app/views/well_known/host_meta/show.xml.ruby:8]: I don't think this is a practical concern.

File: app/services/resolve_account_service.rb (hunk 3 above)
Comment (ThibG) [reply app/services/resolve_account_service.rb:29]: alright

**Review by ClearlyClaire — APPROVED** [review]
I think I would drop the type in the `lrdd` link we return in our own `host-meta`, as we support rendering both `jrd` and `xrd` for that. I'm slightly worried about losing `xrd` handling, but it shouldn't be a big deal.

Finally, a small number of instances might be broken after this change due to misconfiguration (`/.well-known/webfinger` returning a non-404 error or a broken redirection). Ideally, they should be reached out to fix their configuration before a new Mastodon version releases, but ultimately, this is a configuration error on their end.

**Pull request merged**